lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -1341,3 +1388,316 @@ def _capture_libvirt_logs(self) -> None:
+        ls = node.tools[Ls]
+        path = "/dev/vfio/vfio"
+        err = "Host does not support IOMMU"
+        if not ls.path_exists(path=path, sudo=True):
+            raise LisaException(f"{err} : {path} does not exist")
+
+        path = "/sys/kernel/iommu_groups/"
+        if len(ls.list(path=path, sudo=True)) == 0:
+            raise LisaException(f"{err} : {path} does not have any entry")
+
+    def _is_driver_vfio_pci(
+        self,
+        device_addr: DeviceAddressSchema,
+    ) -> bool:
+        lspci = self.host_node.tools[Lspci]
+        device_addr_str = self._get_pci_address_str(device_addr)
+        kernel_module = lspci.get_used_module(device_addr_str)
+        return kernel_module == "vfio-pci"
+
+    def _set_device_passthrough_node_context(
+        self,
+        node_context: NodeContext,
+        node_runbook: BaseLibvirtNodeSchema,
+    ) -> None:
+        if node_runbook.device_passthrough:
+            node_context.is_device_passthrough_set = True
+            for config in node_runbook.device_passthrough:
+                device_context = DevicePassthroughContext()
+                device_context.managed = config.managed
+                device_context.pool_type = config.pool_type
+                devices = self._get_devices_from_pool(config.pool_type, config.count)
+                device_context.device_list = devices
+                node_context.device_passthrough_context.append(device_context)
+
+    def _create_device_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        vendor_id: str,
+        device_id: str,
+    ) -> None:
+        self.AVAILABLE_HOST_DEVICES[pool_type] = {}
+        lspci = self.host_node.tools[Lspci]
+        device_list = lspci.get_devices_by_vendor_device_id(
+            vendor_id=vendor_id,
+            device_id=device_id,
+        )
+        primary_nic_iommu = self._get_primary_nic_id()
+        for item in device_list:
+            device = self.POOL_TYPE_TO_DEVICE_PROPERTY[pool_type]()
+            slot_info = item.slot.split(":")
+
+            device.domain = item.domain
+            device.bus = slot_info[0]
+            device.slot = slot_info[1].split(".")[0]
+            device.function = slot_info[1].split(".")[1]
+            iommu_group = self._get_device_iommu_group(device)
+            is_vfio_pci = self._is_driver_vfio_pci(device)
+
+            if not is_vfio_pci and iommu_group not in primary_nic_iommu:
+                pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, {})
+                devices = pool.get(iommu_group, [])
+                devices.append(device)
+                pool[iommu_group] = devices
+                self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _get_device_iommu_group(self, device: DeviceAddressSchema) -> str:
+        iommu_pattern = re.compile(r"/sys/kernel/iommu_groups/(?P<id>\d+)/devices/.*")
+        device_id = self._get_pci_address_str(device, True)
+        command: str = (
```

Comment:
Don't use shell piping, awk, or grep. Run each command individually and process the output in Python.

Similar comment elsewhere.

Reply:
Sure @cwize1